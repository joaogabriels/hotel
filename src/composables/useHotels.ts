import { ref, computed, watch } from 'vue'
import { useRoute } from 'vue-router'
import { hotelService } from '../services/hotelService'
import type { Hotel } from '../components/models'

export function useHotels() {
  const route = useRoute()
  const hotels = ref<Hotel[]>([])
  const isFetching = ref(false)
  const totalPages = ref(0)
  const currentPage = ref(1)

  const hasNextPage = computed(() => currentPage.value < totalPages.value)

  const fetchHotels = async () => {
    isFetching.value = true

    try {
      const response = await hotelService.getHotels({
        location: route.query.location as string,
        name: route.query.hotelName as string,
        sortBy: route.query.sortBy as 'recommended' | 'best_rated',
        page: Number(route.query.page || 1),
        perPage: 10,
      })

      hotels.value = response.data
      totalPages.value = response.totalPages
      currentPage.value = response.page
    } catch (error) {
      console.error('Error fetching hotels:', error)
      throw error
    } finally {
      isFetching.value = false
    }
  }

  const fetchNextPage = async () => {
    if (!hasNextPage.value || isFetching.value) return

    await fetchHotels()
  }

  watch(
    () => route.query,
    async () => {
      await fetchHotels()
    },
    { immediate: true },
  )

  return {
    hotels,
    isFetching,
    hasNextPage,
    fetchNextPage,
    currentPage,
    totalPages,
  }
}
