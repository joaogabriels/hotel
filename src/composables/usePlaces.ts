import { ref } from 'vue'

import type { Place } from '../components/models'
import { placeService } from '../services/placeService'

export function usePlaces() {
  const places = ref<Place[]>([])
  const isLoading = ref(false)
  const error = ref<Error | null>(null)

  const fetchPlaces = async () => {
    isLoading.value = true
    error.value = null

    try {
      places.value = await placeService.getPlaces()
    } catch (err) {
      error.value = err as Error
      console.error('Error loading places:', err)
    } finally {
      isLoading.value = false
    }
  }

  const filterPlaces = (searchTerm: string) => {
    if (!searchTerm) return places.value

    const needle = searchTerm.toLowerCase()
    return places.value.filter(
      (place) =>
        place.name.toLowerCase().includes(needle) ||
        place.state.name.toLowerCase().includes(needle),
    )
  }

  return {
    places,
    isLoading,
    error,
    fetchPlaces,
    filterPlaces,
  }
}
