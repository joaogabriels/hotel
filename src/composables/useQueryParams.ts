import { useRoute, useRouter } from 'vue-router'

interface QueryParams {
  location?: string
  hotelName?: string
  sortBy?: string
}

export function useQueryParams() {
  const router = useRouter()
  const route = useRoute()

  const updateQueryParams = async (params: Partial<QueryParams>) => {
    const newQuery = {
      ...route.query,
      ...params,
    }

    Object.keys(newQuery).forEach((key) => {
      if (
        newQuery[key as keyof QueryParams] === undefined ||
        newQuery[key as keyof QueryParams] === null ||
        newQuery[key as keyof QueryParams] === ''
      ) {
        delete newQuery[key as keyof QueryParams]
      }
    })

    await router.push({ query: newQuery })
  }

  return {
    updateQueryParams,
  }
}
