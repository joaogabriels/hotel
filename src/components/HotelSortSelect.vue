<template>
  <div class="row items-center text-grey-8">
    <span class="text-body2">Organizar por</span>
    <q-btn-dropdown flat dense no-caps :label="selectedOption?.label" class="q-ml-sm text-primary text-weight-medium">
      <q-list>
        <q-item v-for="option in sortOptions" :key="option.value" clickable v-close-popup @click="handleSelect(option)">
          <q-item-section>
            <q-item-label>{{ option.label }}</q-item-label>
          </q-item-section>
        </q-item>
      </q-list>
    </q-btn-dropdown>
  </div>
</template>

<script setup lang="ts">
import { ref, onMounted } from 'vue';
import { useQueryParams } from '../composables/useQueryParams';
import { useRoute } from 'vue-router';

interface SortOption {
  label: string;
  value: string;
}

const sortOptions: SortOption[] = [
  { label: 'Recomendados', value: 'recommended' },
  { label: 'Melhor avaliados', value: 'best_rated' }
];

const selectedOption = ref(sortOptions[0]);
const { updateQueryParams } = useQueryParams();
const { query } = useRoute();

const handleSelect = async (option: SortOption) => {
  selectedOption.value = option;
  await updateQueryParams({ sortBy: option.value });
};

onMounted(() => {
  if (query.sortBy) {
    selectedOption.value = sortOptions.find(option => option.value === query.sortBy) || sortOptions[0];
  }
});
</script>