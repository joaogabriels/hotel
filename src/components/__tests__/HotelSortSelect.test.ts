import { flushPromises, mount, shallowMount } from '@vue/test-utils'
import type { Mock } from 'vitest'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest'
import { useRoute } from 'vue-router'

import * as useQueryParamsModule from '../../composables/useQueryParams'
import HotelSortSelect from '../HotelSortSelect.vue'
import { QBtnDropdown } from 'quasar'

installQuasarPlugin({ components: { QBtnDropdown } })

vi.mock('vue-router', () => ({
  useRoute: vi.fn(),
}))

describe('HotelSortSelect', () => {
  afterEach(() => {
    vi.clearAllMocks()
  })

  beforeEach(() => {
    vi.spyOn(useQueryParamsModule, 'useQueryParams').mockReturnValue({
      updateQueryParams: vi.fn(),
    })
    ;(useRoute as Mock).mockReturnValue({
      query: {},
    })
  })

  it('matches snapshot', () => {
    const wrapper = shallowMount(HotelSortSelect)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be able to select sorting option', async () => {
    const wrapper = mount(HotelSortSelect)

    await wrapper.vm.handleSelect({ label: 'Melhor avaliados', value: 'best_rated' })

    await flushPromises()

    expect(wrapper.findComponent({ name: 'q-btn-dropdown' }).props('label')).toBe(
      'Melhor avaliados',
    )
  })

  it('should get sorting option from query params on mount', async () => {
    ;(useRoute as Mock).mockReturnValue({
      query: {
        sortBy: 'best_rated',
      },
    })

    const wrapper = mount(HotelSortSelect)

    await flushPromises()

    expect(wrapper.findComponent({ name: 'q-btn-dropdown' }).props('label')).toBe(
      'Melhor avaliados',
    )
  })
})
