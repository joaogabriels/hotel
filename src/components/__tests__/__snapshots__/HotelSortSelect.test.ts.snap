// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HotelSortSelect > matches snapshot 1`] = `
"<div class=\\"row items-center text-grey-8\\"><span class=\\"text-body2\\">Organizar por</span>
  <q-btn-dropdown-stub label=\\"Recomendados\\" replace=\\"false\\" disable=\\"false\\" type=\\"button\\" flat=\\"true\\" outline=\\"false\\" push=\\"false\\" unelevated=\\"false\\" square=\\"false\\" rounded=\\"false\\" glossy=\\"false\\" fab=\\"false\\" fabmini=\\"false\\" nocaps=\\"true\\" nowrap=\\"false\\" dense=\\"true\\" ripple=\\"true\\" align=\\"center\\" stack=\\"false\\" stretch=\\"false\\" transitionshow=\\"fade\\" transitionhide=\\"fade\\" transitionduration=\\"300\\" modelvalue=\\"false\\" split=\\"false\\" cover=\\"false\\" persistent=\\"false\\" noroutedismiss=\\"false\\" autoclose=\\"false\\" menuanchor=\\"bottom end\\" menuself=\\"top end\\" disablemainbtn=\\"false\\" disabledropdown=\\"false\\" noiconanimation=\\"false\\" class=\\"text-primary text-weight-medium\\"></q-btn-dropdown-stub>
</div>"
`;
